balancer/xds/xds.go

```diff
@@ -363,28 +380,69 @@ func (x *xdsBalancer) HandleSubConnStateChange(sc balancer.SubConn, state connec
 }
 
 func (x *xdsBalancer) HandleResolvedAddrs(addrs []resolver.Address, err error) {
-	update := &addressUpdate{
-		addrs: addrs,
-		err:   err,
-	}
-	select {
-	case x.grpcUpdate <- update:
-	case <-x.ctx.Done():
-	}
+	grpclog.Error("UpdateResolverState should be called instead of HandleResolvedAddrs")
+}
+
+type serviceConfig struct {
+	LoadBalancingConfig []*loadBalancingConfig
 }
 
-// TODO: once the API is merged, check whether we need to change the function name/signature here.
-func (x *xdsBalancer) HandleBalancerConfig(config json.RawMessage) error {
-	var cfg xdsConfig
-	if err := json.Unmarshal(config, &cfg); err != nil {
-		return errors.New("unable to unmarshal balancer config into xds config")
+func parseFullServiceConfig(s string) *serviceConfig {
+	var ret serviceConfig
+	err := json.Unmarshal([]byte(s), &ret)
+	if err != nil {
+		return nil
 	}
+	return &ret
+}
+
+func (x *xdsBalancer) UpdateResolverState(s resolver.State) {
+	defer func() {
+		x.lastResolverUpdate = &s
+	}()
 
+	var update interface{}
+	// if service config does not change for this update, we only send address update.
+	if x.lastResolverUpdate != nil && x.lastResolverUpdate.ServiceConfig == s.ServiceConfig {
+		update = &resolverUpdate{
```

Comment:
Should this function just sent a `&resolverUpdate{...}` to the executor goroutine, and do all the checks and parsing there?

Reply:
Why we want to do this?

Reply:
Some of the logic here and later in the goroutine is duplicate.

For example:
Each resolver update, the balancer does:
1. if addresses are updated, do something for the addresses
1. if config is updated, do something for the config

Checks like this are done twice.
First time here,
1. if addresses are updated, `resolverUpdate.addrUpdate` is set
1. if config is updated, `resolverUpdate.xdsconfig` is set

Later in the goroutine
1. if `resolverUpdate.addrUpdate` is not nil, do something
1. if `resolverUpdate.xdsconfig` is not nil, do something

The code is this way because the resolver update is still treated as two separate things. I think the code should be updated to actually treat them as one.

The update sent to the goroutine can be just
```
type resolverUpdate struct {
    state resolver.State
}
```